sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -36,75 +35,155 @@ class GpuGenerateExecSparkPlanMeta(
+   *
+   * This is a specialized method for GPU runtime, which is called by GpuGenerateExec who owns
+   * the generator. The reason of creating a new method rather than implementing columnarEval is
+   * that generator is an integrated Table transformer instead of column transformer in terms of
+   * cuDF.
+   *
+   * @param inputBatch projected input data, which ensures appending columns are ahead of
+   *                   generators' inputs. So, generators can distinguish them with an offset.
+   * @param generatorOffset column offset of generator's input columns in `inputBatch`
+   * @param outer when true, each input row will be output at least once, even if the output of the
+   *              given `generator` is empty.
+   *
+   * @return result ColumnarBatch
+   */
+  def generate(inputBatch: ColumnarBatch,
+    generatorOffset: Int,
+    outer: Boolean): ColumnarBatch
+}
+
+trait GpuCollectionGenerator extends GpuGenerator {
+  /** The position of an element within the collection should also be returned. */
+  def position: Boolean
+
+  /** Rows will be inlined during generation. */
+  def inline: Boolean
+
+  /** The type of the returned collection object. */
+  def collectionType: DataType = dataType
+}
+
+abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuCollectionGenerator {
+  override val inline: Boolean = false
+
+  override def checkInputDataTypes(): TypeCheckResult = child.dataType match {
+    case _: ArrayType | _: MapType =>
+      TypeCheckResult.TypeCheckSuccess
+    case _ =>
+      TypeCheckResult.TypeCheckFailure(
+        "input to function explode should be array or map type, " +
+          s"not ${child.dataType.catalogString}")
+  }
+
+  // hive-compatible default alias for explode function ("col" for array, "key", "value" for map)
+  override def elementSchema: StructType = child.dataType match {
+    case ArrayType(et, containsNull) =>
+      if (position) {
+        new StructType()
+          .add("pos", IntegerType, nullable = false)
+          .add("col", et, containsNull)
+      } else {
+        new StructType()
+          .add("col", et, containsNull)
+      }
+    case MapType(kt, vt, valueContainsNull) =>
+      if (position) {
+        new StructType()
+          .add("pos", IntegerType, nullable = false)
+          .add("key", kt, nullable = false)
+          .add("value", vt, valueContainsNull)
+      } else {
+        new StructType()
+          .add("key", kt, nullable = false)
+          .add("value", vt, valueContainsNull)
+      }
+  }
+
+  override def collectionType: DataType = child.dataType
+}
+
+case class GpuExplode(child: Expression) extends GpuExplodeBase {
+  override val position: Boolean = false
+
+  override def generate(inputBatch: ColumnarBatch,
+    generatorOffset: Int,
+    outer: Boolean): ColumnarBatch = {
+
+    require(inputBatch.numCols() - 1 == generatorOffset,
+      "GpuExplode should has and only has one input attribute")
```

Comment:
Internal Error GpuExplode supports one and only one input attribute.

Reply:
Done.